src/engine/core/FirstPersonCamera.cpp

```diff
@@ -0,0 +1,190 @@
+#include "FirstPersonCamera.hpp"
+
+#include "core/Application.hpp"
+
+namespace lei3d
+{
+
+	FirstPersonCamera::FirstPersonCamera(GLFWwindow* window, float yaw, float pitch, float flySpeed)
+		// this is a member initializer list using uniform initialization C++17 feature
+		: m_Yaw{ yaw }, m_Pitch{ pitch }, m_FlySpeed{ flySpeed }, m_FOVDeg{ 45.0f }, m_NearPlane{ 0.1f }, m_FarPlane{ 700.0f }
+	{
+		m_MouseEnterFlag = true; // the first time the mouse is entered onto the screen, do not flick
+
+		int screenWidth, screenHeight;
+		glfwGetWindowSize(window, &screenWidth, &screenHeight);
+		m_PrevX = screenWidth / 2.0f;
+		m_PrevY = screenHeight / 2.0f;
+		m_Aspect = static_cast<float>(screenWidth) / static_cast<float>(screenHeight);
+
+		m_CameraPos = glm::vec3(0.0f, 0.0f, 3.0f);
+		m_CameraFront = glm::vec3(0.0f, 0.0f, -1.0f);
+		m_CameraUp = glm::vec3(0.0f, 1.0f, 0.0f);
+	}
+
+	FirstPersonCamera::~FirstPersonCamera()
+	{
+		// no need rn
+	}
+
+	void FirstPersonCamera::cameraMouseCallback(double xPosInput, double yPosInput)
+	{
+		float xpos = static_cast<float>(xPosInput);
+		float ypos = static_cast<float>(yPosInput);
+
+		if (m_MouseEnterFlag)
+		{
+			m_PrevX = xpos;
+			m_PrevY = ypos;
+			m_MouseEnterFlag = false;
+		}
+
+		float xoffset = xpos - m_PrevX;
+		float yoffset = m_PrevY - ypos; // openGL inverted y
+		m_PrevX = xpos;
+		m_PrevY = ypos;
+
+		xoffset *= MOUSE_SENSITIVITY;
+		yoffset *= MOUSE_SENSITIVITY;
+
+		m_Pitch += yoffset;
+		m_Yaw += xoffset;
+		RotatePlayer(xoffset);
+
+		if (m_Pitch > MAX_PITCH)
+			m_Pitch = MAX_PITCH;
+		if (m_Pitch < -MAX_PITCH)
+			m_Pitch = -MAX_PITCH;
+
+		glm::vec3 direction;
+		direction.x = cos(glm::radians(m_Yaw)) * cos(glm::radians(m_Pitch));
+		direction.y = sin(glm::radians(m_Pitch));
+		direction.z = sin(glm::radians(m_Yaw)) * cos(glm::radians(m_Pitch));
+		m_CameraFront = glm::normalize(direction) * m_FlySpeed;
+	}
+
+	void FirstPersonCamera::PollCameraMovementInput()
```

Comment:
Don't copy this from FlyCamera. The camera shouldn't be influenced by input (especially the shift *10 thing). It should just follow the player position (or attached entity).

Ideally, we create an FPS camera in the Scene and attach it to an entity so it can just reference the attached entity.